platforms/core-runtime/logging/src/main/java/org/gradle/internal/problems/failure/DeduplicatingFailurePrinter.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.problems.failure;
+
+import javax.annotation.Nullable;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+/**
+ * Formats {@link Failure} instances into stacktraces and keeps track of the unique traces.
+ * <p>
+ * The uniqueness of traces is determined by the N topmost stack frames.
+ */
+public class DeduplicatingFailurePrinter {
```

Comment:
This is a new machinery that will be reused for `LoggingDeprecatedFeatureHandler`.